src/TokenizerFwf.cpp

```diff
@@ -79,12 +79,16 @@ TokenizerFwf::TokenizerFwf(const std::vector<int>& beginOffset, const std::vecto
   cols_(beginOffset.size()),
   moreTokens_(false)
 {
-  if (beginOffset_.size() != endOffset_.size())
+  // File is assumed to be ragged (last column can have variable width)
+  // when length of end vector is one shorter then begin vector.
+  isRagged_ = beginOffset_.size()==(endOffset_.size()+1L);
```

Comment:
I agree that we should only use one single sentinel value (NA).
I think the code will be indeed cleaner if we check for explicit NA inside the cpp code and leave the original length checks on start and end in the R and cpp code.